### PR TITLE
Add TRL GRPO Reasoning with Advanced Reward notebook

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
## Summary
This notebook demonstrates **GRPO (Group Relative Policy Optimization)** fine-tuning for mathematical reasoning using TRL library with advanced reward mechanisms.

## Key Features
- Multi-reward training with 4 different reward functions
- Advanced progress tracking with HuggingFace-style interactive tables
- Mathematical reasoning on GSM8K dataset
- Memory efficient training with 4-bit quantization and LoRA
- Structured output generation with reasoning sections

## Requirements Checklist
- [x] Notebook filename is lowercase
- [x] Added to _toctree.yml
- [x] Added to index.md
- [x] Author attribution included
- [x] Non-informative outputs removed
- [x] No empty code cells
- [x] No custom images requiring upload

@merveenoyan @stevhliu

Contributed by: Behrooz Azarkhalili

### CODE REVIEW

File: notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb [hunk 1]
```diff
@@ -0,0 +1,1004 @@
+{
```
Comment (qgallouedec) [inline notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <p>you can remove torch and accelerate, these are dependencies of trl</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (qgallouedec) [inline notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #33.    if tokenizer.pad_token is None:
</pre><p>this is done automatically, you can remove this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (qgallouedec) [inline notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #13.        num_generations=4,  # Reduce to save memory
</pre><p>reducing num_generation doesn't save memory (it used to be the case, but now, generation and training batches are decoupled). I think you can safely keep the default value (8)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (qgallouedec) [inline notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #23.        remove_unused_columns=False,
</pre><p>this is already false by default, I'd recommend removing for clarity.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (qgallouedec) [inline notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.        processing_class=tokenizer,
</pre><p>you can remove this line, GRPO will load automatically the tokenizer</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: notebooks/en/index.md [hunk 2]
```diff
@@ -7,6 +7,7 @@ applications and solving various machine learning tasks using open-source tools
 
 Check out the recently added notebooks:
 
+- [TRL GRPO Reasoning with Advanced Reward](trl_grpo_reasoning_advanced_reward)
```
Comment (sergiopaniego) [inline notebooks/en/index.md:10]: You can remove the last entry since we aim to have the last 5 here.

File: notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb [hunk 3]
```diff
@@ -0,0 +1,2070 @@
...
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <p>I tested the trick  <code>import trackio as wandb</code> and it seems it did not work. Can you please give me the link to see how it can automatically handle it?</p>

[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: <p>I haven't encountered this issue in the multiple runs I've done. It seems to be resolved now.</p>

[the remark this answers is not in the document]
Comment (sergiopaniego) [reply notebooks/en/trl_grpo_reasoning_advanced_reward.ipynb:1]: This is not yet resolved 😄 We can remove the button from the notebook since it's automatically added afterwards. Could you update it?

**Review by sergiopaniego — COMMENTED** [review]
Just a final comment regarding the removal of Open in Colab button and we're ready!

File: notebooks/en/index.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ applications and solving various machine learning tasks using open-source tools
 Check out the recently added notebooks:
 
 - [Post training an VLM for reasoning with GRPO using TRL](fine_tuning_vlm_grpo_trl)
+- [TRL GRPO Reasoning with Advanced Reward](trl_grpo_reasoning_advanced_reward)
 - [Fine-Tuning a Vision Language Model with TRL using MPO](fine_tuning_vlm_mpo)
 - [Fine tuning a VLM for Object Detection Grounding using TRL](fine_tuning_vlm_object_detection_grounding)
 - [Hyperparameter Optimization with Optuna and Transformers](optuna_hpo_with_transformers)
```
Comment (sergiopaniego) [inline notebooks/en/index.md:14]:
```suggestion
```

**Review by sergiopaniego — APPROVED** [review]
thanks!!
